ocaml-lsp-server/src/fmt.ml

```diff
@@ -1,40 +1,44 @@
 open Import
 
-let read_to_end (in_chan : in_channel) : string =
-  let buf = Buffer.create 0 in
-  let chunk_size = 1024 in
-  let chunk = Bytes.create chunk_size in
-  let rec go pos =
-    let actual_len = input in_chan chunk pos chunk_size in
-    if actual_len > 0 then (
-      Buffer.add_subbytes buf chunk 0 actual_len;
-      go pos
-    )
-  in
-  go 0;
-  Buffer.contents buf
-
 type command_result =
   { stdout : string
   ; stderr : string
   ; status : Unix.process_status
   }
 
-let run_command command stdin_value args : command_result =
-  let command =
-    match args with
-    | [] -> command
-    | _ -> Printf.sprintf "%s %s" command (String.concat ~sep:" " args)
+let run_command scheduler thread bin stdin_value args : command_result Fiber.t =
+  let open Fiber.O in
+  let stdin_i, stdin_o = Unix.pipe ~cloexec:true () in
+  let stdout_i, stdout_o = Unix.pipe ~cloexec:true () in
+  let stderr_i, stderr_o = Unix.pipe ~cloexec:true () in
+  let pid =
+    let args = Array.of_list (bin :: args) in
+    Unix.create_process bin args stdin_i stdout_o stderr_o |> Stdune.Pid.of_int
+  in
+  Unix.close stdin_i;
+  Unix.close stdout_o;
+  Unix.close stderr_o;
+  let task =
+    Scheduler.async_exn thread (fun () ->
+        let out_chan = Unix.out_channel_of_descr stdin_o in
+        output_string out_chan stdin_value;
+        flush out_chan;
+        close_out out_chan;
+        let stdout_in = Unix.in_channel_of_descr stdout_i in
+        let stderr_in = Unix.in_channel_of_descr stderr_i in
+        let stdout = Stdune.Io.read_all stdout_in in
+        close_in_noerr stdout_in;
+        let stderr = Stdune.Io.read_all stderr_in in
+        close_in_noerr stderr_in;
+        (stdout, stderr))
+  in
+  let+ status, (stdout, stderr) =
+    Fiber.fork_and_join
+      (fun () -> Scheduler.wait_for_process scheduler pid)
+      (fun () ->
+        let+ res = Scheduler.await_no_cancel task in
```

Comment:
Perhaps `await_no_cancel` could be modified to return the exit status? If figure you want it often enough and it would save the cost of spawning an additional thread.

Reply:
> Perhaps await_no_cancel could be modified to return the exit status

It cannot be. All the waiting for processes to finish is done in a dedicated thread in the scheduler.

> If figure you want it often enough and it would save the cost of spawning an additional thread.

Note that neither of these fibers create a thread. The first fiber asks the process watcher thread to watch for the pid, while the second fiber is waiting for a task enqueued in an existing thread to finish.